### ISSUE TITLE
Training issue with ram caching & DataLoader

### ISSUE THREAD

**Issue by wvalcke** [ultralytics/ultralytics#1010]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Training

### Bug

With 
* Ultralytics YOLOv8.0.38 🚀 Python-3.10.6 torch-1.13.1+cu117 CUDA:0 
* Linux mint 21.1
* Python 3.10.6
i discovered a very strange (and reproducible behavior).
When training (on single GPU) a network yolov8s i get the following (very good) results (trained for 1200 epochs)

![image](https://user-images.githubusercontent.com/22256774/219372533-fc69b512-48fb-4fce-94de-24eb99ff6c01.png)
and losses look like 
![image](https://user-images.githubusercontent.com/22256774/219372711-9b833a7e-55e7-4b2c-888b-d4aaed2d7a54.png)
Now when retraining with only the following changes:
* cache=True (caching image dataset to ram)
* workers=1 (Need to set workers to 1, otherwise the dataloader hangs when caching to ram is active)
The training now gives as results (stopped training after 500 epocs with CTRL-C)

![image](https://user-images.githubusercontent.com/22256774/219373429-7060e38c-88ba-47e1-805e-9018d53a5550.png)
And the loss looks like
![image](https://user-images.githubusercontent.com/22256774/219373711-c042f335-5c66-4ca9-be74-c6215aa72b6e.png)

This behavior is always reproducible, and there must be a problem with the ram cache & dataloader. It was reported before already that we have issues with caching to ram and workers > 1 (see issue #673) and also issue #957. 

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@wvalcke thanks for the bug report. All official models were trained with cache=disk, so you might use that as a patch while we investigate.

**Comment by glenn-jocher** [comment]
@wvalcke I'm not able to reproduce this. I trained 3 models (cache=False, cache=disk, cache=ram) on our A100 server in our Docker image and they all returned near identical results, and I saw the same with 2 additional DDP runs. So caching is working correctly.


sleep 0 && yolo task=detect mode=train device=0 model=yolov8n.pt data=coco.yaml epochs=50 batch=128 project=test-cache name=cache-None cache=None > /dev/null 2>&1 &
sleep 2 && yolo task=detect mode=train device=1 model=yolov8n.pt data=coco.yaml epochs=50 batch=128 project=test-cache name=cache-disk cache=disk > /dev/null 2>&1 &
sleep 4 && yolo task=detect mode=train device=2 model=yolov8n.pt data=coco.yaml epochs=50 batch=128 project=test-cache name=cache-ram cache=ram > /dev/null 2>&1 &

sleep 6 && yolo task=detect mode=train device=3,4 model=yolov8n.pt data=coco.yaml epochs=50 batch=128 project=test-cache name=cache-disk-DDP cache=disk > /dev/null 2>&1 &
sleep 8 && yolo task=detect mode=train device=5,6 model=yolov8n.pt data=coco.yaml epochs=50 batch=128 project=test-cache name=cache-ram-DDP cache=ram > /dev/null 2>&1 &

**Comment by wvalcke** [comment]
Hi Glenn,

Training for 50 epochs is too little. The problem occurs when training for at least 400-500 epochs. To avoid very long runs, the training is done on a toy set with limited images. 
You can see the graphs above, you can train for 1200 epochs on the data set without problem (mosaic training active). But not with cache=ram, then the problem occurs after 400 epochs.
If you want i can prepare a dataset which produces the problem, then you must be able to reproduce the problem. 
Also keep in mind that some people complained about a locked up training when cache=ram and workers=8. There must be something wrong, very likely it's threading related.

**Comment by glenn-jocher** [comment]
@wvalcke the 50 epoch training I did was on COCO at batch=128, so 1000 batches per epoch, i.e. 50,000 iterations total. There was about zero deviation among any of the trained models.

Furthermore all official models are trained to 500 epochs using cache=disk, and the results are what we show in the README.

**Comment by wvalcke** [comment]
Hi @glenn-jocher ,

I just tried again with my simple dataset, i can easily reproduce the error. If you are interested in the data set, i'll prepare it so you can see yourself. The training time takes 20 minutes for 1200 epochs, using nano v8 model. (using HP Z640 workstation, 1x RTX 2080Ti)
Thanks.
